Sort it minigame

RedstoneMedia opened the pull request:
This PR adds a game, about sorting boxes, that fall from the sky, faster than your opponents. \
It is a 3D local multiplayer game, that can be played with up to 4 players (With a keyboard and controllers).
In that sense it is inspired by Wii-Party mini-games.
![image](https://user-images.githubusercontent.com/34373974/172901769-6eb65659-0ce1-4b9d-901f-db839184b7bc.png)

RedstoneMedia committed:
Added sort it game (Movement, Boxes)

Added split screen + working pedestals

Added start UI + seperate player controlls

Added Icon; Improved pedestal attract; Added timer

Refactored scene names

Added maximum box count

Added premature game end

Applied irrational code style guidelines (Like seriously. Maximum line lengths on comments ???)

Fixed example game

Numenter commented:
Don't commit .blend1 files, they are Blender backup files.

Use sanke_case for file names.
https://docs.godotengine.org/en/stable/tutorials/scripting/gdscript/gdscript_styleguide.html#naming-conventions

Numenter commented:
Don't commit .blend1 files, they are Blender backup files.

Use snake_case for file names.
https://docs.godotengine.org/en/stable/tutorials/scripting/gdscript/gdscript_styleguide.html#naming-conventions

RedstoneMedia commented:
I am pretty sure, that that only applies to scripts, not Scenes. \
The page you linked is the "gdscript_styleguide". Scenes are not written in GDScript.

RedstoneMedia committed:
Removed .blend1 files

Numenter commented:
Consistency is key. All files should follow the same naming convention.
_(that's my opinion :P)_

RedstoneMedia commented:
Well the way I did it is consistent. I always name Scene files in CamelCase and script files in snake_case. That way I can more easily distinguish between a Scene and the attached script of that scene. *(also my opinion)*

RedstoneMedia committed:
Organized assests; Added floor shader

Joshix-1 commented:
1. The game doesn't load correctly, some you specified some file names in the incorrectly
2. I don't understand the game

Numenter commented:
- Warning: _open: Case mismatch opening requested file 'StatusDisplay.tscn', stored as 'statusDisplay.tscn' in the filesystem. This file will not open when exported to other case-sensitive platforms.
- Some material names have spaces. (Could be a problem, i dont know)

RedstoneMedia committed:
Renamed sort it scene files

RedstoneMedia commented:
Yeah, sorry my bad. I just now set `core.ignorecase` to false, so I missed this issue on Windows, since I had already renamed the files locally.

ASecondGuy commented:
Gameplay:
- It is unclear how you get points.
- The Boxes should be labeled on all sides.
- It is hard to get boxes on the pedestals.

Code:
- sometimes you have onready vars but don't use them.
- the Pedestals aren't on 0,0 in their scene.
- make sure to use onready or local vars when you need a node more than once.

That's everything I got for now.

RedstoneMedia commented:
I'm currently working on a small tutorial page.

I planned to get the numbers on all sides (That's why there is more than just an "Up" node), but I got stuck, since I had no idea how to calculate the right rotation using the anchor points as normal vectors from the node origin (still don't really know how).

The pedestals are purposefully not at 0,0 In the scene, because the map is not perfectly symmetrical either and because the players always spawn on the right side of the pedestals, meaning that the players on the right side would have less space to the right, if they were at 0,0 (because of the map).

Would you mind elaboration on what exactly is too hard about getting the boxes on the pedestals ?
Is it the player controller or the size of the pedestals (their attract range is actually quite high) ? (Do keep in mind that the game is supposed to be *a bit* hard, but it obviously shouldn't be impossible for the average gamer)

Anyways: Thanks for actual feedback on the game.

ASecondGuy commented:
It is hard to get the cubes on the pedestal because the controls don't allow precise enough turning.
When I turn the cube moves from one side of the pedestal to the other in a single frame. Mostly the same when I walk but not as bad.

RedstoneMedia committed:
New Help popup + Control info;
Refactored onready

RedstoneMedia committed:
Merge branch 'main' into sort_it

RedstoneMedia committed:
Kinda Improved player controls; Changed max number

RedstoneMedia committed:
Merge branch 'main' into sort_it

RedstoneMedia committed:
Renamed Box class name; Alternate control type

RedstoneMedia commented:
In the latest commit I added a different control type (Needs to be enabled in the Player scene → "Experimental Rotational Controls"). This is based on the imprecise turning problem, but I am not sure if it resolves this problem or just makes it even harder to move around. That's why I would appreciate **feedback** on this.

Honestly I'm kinda stuck on the character controller, since most 3D top down games use a mouse for turning, but this game obviously can't (You would need to connect one mouse for every player).

RedstoneMedia committed:
Merge branch 'letsgamedev:main' into sort_it

RedstoneMedia committed:
Added dot to confuseable numbers (69)

Switched player controler

Improved character rotation + More obstacles

Numbers are now on all box sides

ASecondGuy commented:
The movement is way better than last time. 
It is still unclear how the numbers on the boxes relate to the points they give.
Sometimes I got lost so maybe add a minimap or an arrow that always points toward the pedestals.
The player also can't look very far so the search for boxes is mostly aimless walking. Maybe change the perspective a bit?
And sometimes the walls block your view so maybe do something about that.

ASecondGuy commented:
I played it again and now I understand how you get points.
I also noticed the arrows on the pedestals. Those are way to small and not clear enough.
Maybe use big > or < between the pedestals. You could even mix them up to make it more interesting.

RedstoneMedia committed:
Increased fov

Numenter commented on game/games/sortit/player.gd:
```diff
@@ -0,0 +1,184 @@
+	_main_body_mesh.mesh.surface_set_material(0, material)
+
+
+func _input(_event):
```
The bug:
- Its not possible to to drop both cubes at the same time.
1. _players.is_action_just_pressed is somehow called on release
2. q and e are somehow connected (try hold q and press e)
3. no InputEventKey echo handling

I tried to fix it, but I gave up after an hour.

RedstoneMedia replied:
Some of this is important, but the chance, that a player presses two buttons in the exact same frame is quite low.

RedstoneMedia replied:
Also, why would you hold down a button ? The magnets are toggled

Numenter replied:
You miss the point. Holding the button shows that there is a connection between q and e.
And it don't have to be in the same frame.

RedstoneMedia replied:
Ok, maybe I phrased my response a bit confusingly. My argument is: I never once got this bug while actually playing the game (Not looking for bugs). Because I never pressed the buttons in that order and time while playing. Secondly, I have already tried to fix this bug, but I got stuck too. I do agree that this is not optimal, but currently I don't have the capacity to fix a bug, that does not occur that often. I will probably try to fix this again, but I will probably do this in a second PR later.

Numenter replied:
found it https://github.com/RedstoneMedia/Suffragium/pull/1

RedstoneMedia replied:
?

RedstoneMedia replied:
Nice

Numenter reviewed (changes requested):
Input system needs some love.

Numenter committed:
input fix

RedstoneMedia committed:
Merge pull request #1 from Numenter/sortit-input-fix

Input fix

RedstoneMedia committed:
Added big arrow to pedestal

RedstoneMedia committed:
Merge branch 'letsgamedev:main' into sort_it

RedstoneMedia committed:
Fixed controller input

Added pointing arrow

Improved mini map

RedstoneMedia committed:
Merge branch 'sort_it' of https://github.com/RedstoneMedia/Suffragium into sort_it
